### PR TITLE
Backport: opt out for Navigation fallback creation

### PR DESCRIPTION
**Pull request by getdave**
Updates original backport from https://core.trac.wordpress.org/ticket/58557 to also allow developers to opt out of the auto-creation of the fallback.

Why? Because if we automatically create something (even if it requires interaction with the WP site or editor) we should allow folks to opt out of that.

Trac ticket: https://core.trac.wordpress.org/ticket/58750

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/class-wp-navigation-fallback.php [hunk 1]
```diff
@@ -26,9 +26,18 @@ class WP_Navigation_Fallback {
...
 	public static function get_fallback() {
 
+		/**
+		 * Filters whether or not a fallback should be created.
+		 *
+		 * @since 6.3.0
+		 *
+		 * @param bool Whether or not to create a fallback.
+		 */
+		$should_create_fallback = apply_filters( 'wp_navigation_should_create_fallback', true );
```
Comment (getdave) [inline src/wp-includes/class-wp-navigation-fallback.php:36]: I'm unsure as to whether the `wp_` prefix is correct.

In the original Gutenberg PR this was prefixed with `gutenberg_`.

File: src/wp-includes/class-wp-navigation-fallback.php [hunk 2]
```diff
@@ -26,9 +26,18 @@ class WP_Navigation_Fallback {
...
+
 		$fallback = static::get_most_recently_published_navigation();
 
-		if ( $fallback ) {
+		if ( $fallback || ! $should_create_fallback ) {
```
Comment (getdave) [inline src/wp-includes/class-wp-navigation-fallback.php:40]: Note: we're not opting out of fetching a fallback, it's opting out of creating one either from:

- converting a classic menu
- creating a new block based menu

File: tests/phpunit/tests/editor/navigation-fallback.php [hunk 3]
```diff
@@ -56,6 +56,25 @@ public function test_should_return_a_default_fallback_navigation_menu_in_absence
 		$this->assertCount( 1, $navs_in_db, 'The fallback Navigation post should be the only one in the database.' );
 	}
 
+	/**
+	 * @ticket 58557
```
Comment (getdave) [inline tests/phpunit/tests/editor/navigation-fallback.php]:
```suggestion
	 * @ticket 58750
```

I believe this needs to reference https://core.trac.wordpress.org/ticket/58750 ?

File: tests/phpunit/tests/editor/navigation-fallback.php [hunk 4]
```diff
@@ -56,6 +56,25 @@ public function test_should_return_a_default_fallback_navigation_menu_in_absence
...
+	 * @covers WP_REST_Navigation_Fallback_Controller::get_fallback
+	 */
+	public function test_should_not_automatically_create_fallback_if_filter_is_falsey() {
+
+		add_filter( 'gutenberg_navigation_should_create_fallback', '__return_false' );
```
Comment (getdave) [inline tests/phpunit/tests/editor/navigation-fallback.php]:
```suggestion
		add_filter( 'wp_navigation_should_create_fallback', '__return_false' );
```

File: tests/phpunit/tests/editor/navigation-fallback.php [hunk 5]
```diff
@@ -56,6 +56,25 @@ public function test_should_return_a_default_fallback_navigation_menu_in_absence
...
+
+		$data = Gutenberg_Navigation_Fallback::get_fallback();
+
+		$this->assertEmpty( $data );
+
+		$navs_in_db = $this->get_navigations_in_database();
+
+		$this->assertCount( 0, $navs_in_db, 'The fallback Navigation post should not have been created.' );
+
+		remove_filter( 'gutenberg_navigation_should_create_fallback', '__return_false' );
```
Comment (getdave) [inline tests/phpunit/tests/editor/navigation-fallback.php]:
```suggestion
		remove_filter( 'wp_navigation_should_create_fallback', '__return_false' );
```

File: tests/phpunit/tests/editor/navigation-fallback.php [hunk 6]
```diff
@@ -56,6 +56,25 @@ public function test_should_return_a_default_fallback_navigation_menu_in_absence
...
 
+	/**
+	 * @ticket 58750
+	 * @covers WP_REST_Navigation_Fallback_Controller::get_fallback
+	 */
+	public function test_should_not_automatically_create_fallback_if_filter_is_falsey() {
+
+		add_filter( 'wp_navigation_should_create_fallback', '__return_false' );
+
+		$data = Gutenberg_Navigation_Fallback::get_fallback();
```
Comment (spacedmonkey) [inline tests/phpunit/tests/editor/navigation-fallback.php]:
```suggestion
		$data = WP_Navigation_Fallback::get_fallback();
```

File: src/wp-includes/class-wp-navigation-fallback.php (hunk 1 above)
Comment (ramonjd) [reply src/wp-includes/class-wp-navigation-fallback.php:36]: The `wp_` prefix is good. 👍🏻 

`gutenberg_*` shouldn't be in core files.

Any third parties who want to use this filter will forever be using `wp_navigation_should_create_fallback` 😄

File: src/wp-includes/class-wp-navigation-fallback.php [hunk 7]
```diff
@@ -26,9 +26,18 @@ class WP_Navigation_Fallback {
 	 */
 	public static function get_fallback() {
 
+		/**
+		 * Filters whether or not a fallback should be created.
+		 *
+		 * @since 6.3.0
+		 *
+		 * @param bool Whether or not to create a fallback.
```
Comment (ramonjd) [inline src/wp-includes/class-wp-navigation-fallback.php]:
```suggestion
		 * @param bool Whether to create a fallback navigation menu. Default true.
```

File: tests/phpunit/tests/editor/navigation-fallback.php [hunk 8]
```diff
@@ -56,6 +56,25 @@ public function test_should_return_a_default_fallback_navigation_menu_in_absence
 		$this->assertCount( 1, $navs_in_db, 'The fallback Navigation post should be the only one in the database.' );
 	}
 
+	/**
+	 * @ticket 58750
```
Comment (ramonjd) [inline tests/phpunit/tests/editor/navigation-fallback.php:60]:
```suggestion
	/**
	 * @ticket 58750
	 *
```

**Review by ramonjd — APPROVED** [review]
LGTM to me. 
A couple of nitpicks.

**Review by spacedmonkey — APPROVED** [review]

**Comment by tellthemachines** [comment]
Committed in [r56202](https://core.trac.wordpress.org/changeset/56202) / [a02886a](https://github.com/WordPress/wordpress-develop/commit/a02886a868e4bdeb586658926607bdffdd65a7f2).

**Pull request closed without merging**